core/vm/evmc.go

```diff
@@ -0,0 +1,376 @@
+		isHomestead := env.ChainConfig().IsHomestead(env.BlockNumber)
+		if !isHomestead && err == ErrCodeStoreOutOfGas {
+			err = nil
+		}
+		if err == errExecutionReverted {
+			// Assign return buffer from REVERT.
+			// TODO: Bad API design: return data buffer and the code is returned in the same place. In worst case
+			//       the code is returned also when there is not enough funds to deploy the code.
+			output = createOutput
+		}
+	case evmc.Create2:
+		var createOutput []byte
+		createOutput, createAddr, gasLeftU, err = env.Create2(host.contract, input, gasU, value, salt)
+		if err == errExecutionReverted {
+			// Assign return buffer from REVERT.
+			// TODO: Bad API design: return data buffer and the code is returned in the same place. In worst case
+			//       the code is returned also when there is not enough funds to deploy the code.
+			output = createOutput
+		}
+	default:
+		panic(fmt.Errorf("EVMC: Unknown call kind %d", kind))
+	}
+
+	// Map errors.
+	if err == errExecutionReverted {
+		err = evmc.Revert
+	} else if err != nil {
+		err = evmc.Failure
+	}
+
+	gasLeft = int64(gasLeftU)
+	return output, gasLeft, createAddr, err
+}
+
+func getRevision(env *EVM) evmc.Revision {
+	n := env.BlockNumber
+	conf := env.ChainConfig()
+	if conf.IsPetersburg(n) {
+		return evmc.Constantinople2
+	}
+	if conf.IsConstantinople(n) {
+		return evmc.Constantinople
+	}
+	if conf.IsByzantium(n) {
+		return evmc.Byzantium
+	}
+	if conf.IsEIP158(n) {
+		return evmc.SpuriousDragon
+	}
+	if conf.IsEIP150(n) {
+		return evmc.TangerineWhistle
+	}
+	if conf.IsHomestead(n) {
+		return evmc.Homestead
+	}
+	return evmc.Frontier
+}
+
+func (evm *EVMC) Run(contract *Contract, input []byte, readOnly bool) (ret []byte, err error) {
+	evm.env.depth++
+	defer func() { evm.env.depth-- }()
+
+	// Don't bother with the execution if there's no code.
+	if len(contract.Code) == 0 {
+		return nil, nil
+	}
+
+	kind := evmc.Call
+	if evm.env.StateDB.GetCodeSize(contract.Address()) == 0 {
+		// Guess if this is a CREATE.
+		kind = evmc.Create
+	}
+
+	// Make sure the readOnly is only set if we aren't in readOnly yet.
+	// This makes also sure that the readOnly flag isn't removed for child calls.
+	if readOnly && !evm.readOnly {
+		evm.readOnly = true
+		defer func() { evm.readOnly = false }()
+	}
+
+	output, gasLeft, err := evm.instance.Execute(
+		&HostContext{evm.env, contract},
+		getRevision(evm.env),
+		kind,
+		evm.readOnly,
+		evm.env.depth-1,
+		int64(contract.Gas),
+		contract.Address(),
+		contract.Caller(),
+		input,
+		common.BigToHash(contract.value),
+		contract.Code,
+		common.Hash{})
+
+	contract.Gas = uint64(gasLeft)
+
+	if err == evmc.Revert {
+		err = errExecutionReverted
+	} else if evmcError, ok := err.(evmc.Error); ok && evmcError.IsInternalError() {
+		panic(fmt.Sprintf("EVMC VM internal error: %s", evmcError.Error()))
+	}
+
+	return output, err
+}
+
+func (evm *EVMC) CanRun(code []byte) bool {
```

Comment:
This code segment will hijack EVM too, even if I only want to configure EVMC for eWASM. Please make the EVM specializable to either eWASM or EVM, and hard select one or the other based on what it was constructed for.

Reply:
Firstly, in practice you can still control this by EVMC options. E.g. for Hera you can set something like `evm=off` so it will not report the EVM capability.

Secondly, I'm not sure what exactly are you asking for. Something like
```
func NewEVMC(ewasm bool, ...)

NewEVMC(false, ...)
NewEVMC(true, ...)
```

If yes, that sounds like a good change.

Reply:
My understanding of what Peter wants to say, and to summarize grossly, is that we want to have the equivalent of `evm=on`. EVMC taking over EVM needs to be a conscious choice by the user, not the user.

Here, the default behavior of an EVMC backend client is to grab EVM code as soon as `--vm.ewasm='libhera.so'` is set. We want the request for `CapabilityEVM1` to only occur if the client expressedly said `--vm.evm='libhera.so'` or something similar on the command line.

Reply:
> This code segment will hijack EVM too, even if I only want to configure EVMC for eWASM. Please make the EVM specializable to either eWASM or EVM, and hard select one or the other based on what it was constructed for.

This has been addressed. Now we have independent module instances for EVM and Ewasm.